### PR TITLE
Add U-Net

### PR DESCRIPTION
**Pull request by xin-huang**
## Summary by Sourcery

Introduce a UNet-based genotype-matrix training pipeline alongside existing feature-vector workflows, adding configuration, simulation, preprocessing, IO, and model support for windowed HDF5 datasets.

New Features:
- Add UNet and UNet+RNN segmentation models with HDF5-based training and inference support, including custom dataloaders and residual UNet++ layers.
- Introduce a genotype-matrix simulation and preprocessing pipeline that generates fixed-size genotype matrices from msprime simulations and VCFs, with optional sorting and HDF5/TSV output.
- Add binary per-allele labeling via a new BinaryAlleleLabeler for training site-level models on true introgression tracts.
- Extend configuration schemas, CLI train/infer flows, and tests to distinguish feature-vector vs genotype-matrix workflows and to support the new UNet model type.

Enhancements:
- Refactor window generation and sample-name handling into reusable utilities, and add robust HDF5 IO helpers for initializing and appending unified-schema training/inference datasets.
- Improve multiprocessing manager behavior, simulation/preprocessing APIs, and test coverage to better support parallel UNet training data generation.
- Tighten Pydantic configs by forbidding extra fields, normalizing paths, and splitting simulation/preprocess configs into type-specific variants.

Tests:
- Add comprehensive unit and integration tests for UNet layers and model wrapper, HDF5 IO utilities, dataloaders, genotype-matrix simulator/preprocessor, new labeler, and updated configs, simulation, preprocessing, and CLI flows.

### CODE REVIEW

File: gaishi/preprocessors/genotype_matrix_preprocessor.py [hunk 1]
```diff
@@ -17,10 +18,260 @@
...
+                tgt_gts, tgt_samples, self.tgt_rdm_spl_idx
+            )
+            ref_gts, ref_samples = self._sort_ref_genotypes(
+                ref_gts, tgt_gts, ref_samples, self.ref_rdm_spl_idx
+            )
+        else:
+            for i in self.ref_rdm_spl_idx:
+                ref_samples.append(ref_samples[i])
+            for i in self.tgt_rdm_spl_idx:
+                tgt_samples.append(tgt_samples[i])
```
Comment (xin-huang) [inline gaishi/preprocessors/genotype_matrix_preprocessor.py:141]: should not append additional random sample list in this module

File: gaishi/preprocessors/genotype_matrix_preprocessor.py [hunk 2]
```diff
@@ -17,10 +18,260 @@
...
+        metric : str, optional
+            Distance metric to use. Default: 'euclidean'.
+
+        Returns
+        -------
+        tuple[np.ndarray, list]
+            A tuple containing the sorted reference genotypes array and the reordered list of sample identifiers.
+        """
+        for i in rdm_spl_idx:
+            samples.append(samples[i])
```
Comment (xin-huang) [inline gaishi/preprocessors/genotype_matrix_preprocessor.py:195]: should not append additional random sample list in this module

File: gaishi/preprocessors/genotype_matrix_preprocessor.py [hunk 3]
```diff
@@ -17,10 +18,260 @@
...
+        metric : str, optional
+            Distance metric to use. Default: 'euclidean'.
+
+        Returns
+        -------
+        tuple[np.ndarray, list]
+            A tuple containing the sorted target genotypes array and the reordered list of sample identifiers.
+        """
+        for i in rdm_spl_idx:
+            samples.append(samples[i])
```
Comment (xin-huang) [inline gaishi/preprocessors/genotype_matrix_preprocessor.py:232]: should not append additional random sample list in this module

File: gaishi/models/unet/unet_layers.py [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+    output_head : torch.nn.Conv2d
+        Final ``1x1`` convolution mapping features to logits.
+    """
+
+    def __init__(self, num_classes: int, input_channels: int = 3):
+        super().__init__()
+
+        channel_dims = [32, 64, 128, 256, 512]
+
+        self.downsample = nn.MaxPool2d(kernel_size=2, stride=2)
```
Comment (xin-huang) [inline gaishi/models/unet/unet_layers.py]: downsample both variants and samples

File: gaishi/models/unet/unet_layers.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+        gru_layers: int = 1,
+        bidirectional: bool = True,
+    ):
+        super().__init__()
+
+        self.polymorphisms = polymorphisms
+        self.hidden_dim = hidden_dim
+        self.gru_layers = gru_layers
+        self.bidirectional = bidirectional
+
```
Comment (xin-huang) [inline gaishi/models/unet/unet_layers.py]:
```
directions = 2 if bidirectional else 1
self.gru_proj = nn.Linear(directions * hidden_dim, 1)
```

File: gaishi/models/unet/unet_layers.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+        gru_input = torch.cat([unet_rows, gap_rows], dim=-1)  # (B*H, W, 3)
+
+        gru_output = self.gru(gru_input)[0]  # (B*H, W, hidden_dim * directions)
+
+        directions = 2 if self.bidirectional else 1
+        gru_output = gru_output.view(b * h, w, directions, self.hidden_dim)
+
+        # Keep the last hidden unit per direction; average directions -> (B*H, W)
+        gru_scalar = gru_output[..., -1].mean(dim=2)
+
```
Comment (xin-huang) [inline gaishi/models/unet/unet_layers.py]:
```
gru_output, _ = self.gru(gru_input)              # (B*H, W, directions*hidden_dim)
gru_scalar = self.gru_proj(gru_output).squeeze(-1)  # (B*H, W)
fused = gru_scalar + unet_rows.squeeze(-1)
```

File: gaishi/simulators/genotype_matrix_simulator.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        )
+
+        data_dict = preprocessor.run(
+            **list(polymorphism_data_generator.get())[0],
+        )[0]
+
+        if rep is not None:
+            data_dict["Replicate"] = rep
+        if seed is not None:
+            data_dict["Seed"] = seed
```
Comment (xin-huang) [inline gaishi/simulators/genotype_matrix_simulator.py:235]: not in h5 file

**Comment by xin-huang** [comment]
@sourcery-ai

**Comment by xin-huang** [comment]
@sourcery-ai
